### ISSUE TITLE
GMRES Fails Silently and Frequently from Stagnation

### ISSUE THREAD

**Issue by RaderJason** [google/jax#15837]
### Description

`jax.scipy.sparse.linalg.gmres` often fails to converge when `restart` is less than the problem dimension, but will return without indicating failure. This happens in both batch and incremental mode (the below code only tests batch.)

This failure (likely due to stagnation) seems to depend on the absolute difference between the problem dimension and the Krylov subspace dimension (`restart`.) ie. roughly 90% of the time GMRES will fail silently on a problem of dimension 25 if `restart` = 20, and will still fail 90% of the time with a problem of dimension 100 and `restart` = 95. The failure rate jumps as high as 30% with even a 1 dimensional difference between the problem and Krylov subspace.

In addition to the dangerous silent failure, this has some major implications for the implementation of the algorithm as a whole. It implies 30% of the time an incremental solve with `restart >= problem dim` may require computing the Kyrlov subpsace all the way to `problem_dim`, and 90% of the time to within 5 dimensions of it.

Reproducing code:
```
import jax
import jax.numpy as jnp 
import jax.random as jr
import jax.scipy as jsp

import random


def randomkey():
    return jr.PRNGKey(random.randint(0, 2**31 - 1))


# very loose tolerance
tol = 1
absolute_difference = (0, 1, 5)
problem_dims = (25, 100)
for dim in problem_dims:
    for excess_dim in absolute_difference:
        succeeded = 0
        # the failure rate is dependent upon the asbolute difference between
        # the dimension of the problem `dim` and the size of the Krylov subspace
        # `restart`. ie. the failure rate depends on `excess_dim` only.
        restart = dim - excess_dim
        for _ in range(100):
            # force an acceptable condition number. Lower condition numbers will exhibit
            # the same phenomenon.
            cond = jnp.inf
            while cond > 1000:
                matrix = jr.normal(randomkey(), (dim, dim))
                cond = jnp.linalg.cond(matrix)

            true_vector = jr.normal(randomkey(), (dim,))
            b = matrix @ true_vector

            gmres_jit = jax.jit(jsp.sparse.linalg.gmres, static_argnames="restart")
            jax_solution, info = gmres_jit(matrix, b, restart=restart)

            # info == 0 when jax.scipy believes there was no failure.
            assert jnp.all(info == 0.0)

            succeeded = succeeded + (
                jnp.sqrt(jnp.sum((jax_solution - true_vector) ** 2)) < tol
            )

        print(
            f"Number which succeeded for dim = {dim}"
            f" and restart = {restart}: {succeeded}"
        )
@RaderJason

```

### What jax/jaxlib version are you using?

jax v0.4.6, jaxlib v0.4.6

### Which accelerator(s) are you using?

CPU

### Additional system info

python 3.10.9

**Comment by RaderJason** [comment]
So, there's a couple bits I've noticed here:
1. The floating point precision makes no difference for this issue.
2. Scipy also fails in the test above but detects this failure and reports it via info, but info in JAX doesn't return this information. This seems to have been brought up before https://github.com/google/jax/issues/7909 but info still only signals failure when the output is nan, meaning it doesn't capture this failure mode.
3. As an addition to 2., BiCGStab still doesn't have any info! It doesn't capture **any** failure mode. This could be (should be?) risen as its own, separate issue.
4. GMRES(k) has two big failure modes: breakdown and stagnation. For dense matrices as above, we will usually encounter a breakdown. Stagnation seems to be the more common issue in practice though. Theoretically, JAX's GMRES has a check for breakdown, but this check is `v_norm_1 == 0`, which we almost never hit in floating point arithmetic. Really this should be `v_norm_1 < tol` for an appropriate tolerance. (Also, breakdown will stop the loop but isn't reported anywhere.)
5. There is no check for stagnation. Admittedly, this is not standard and I haven't seen an autodifferentiable implementation of GMRES(k) which has a check for this. If the goal is to match Scipy, this could be omitted.

**Comment by shoyer** [comment]
The goal is definitely to match SciPy. So if possible, it would be great to accurately report info! PRs to improve the situation would be gratefully accepted here.

**Comment by patrick-kidger** [comment]
For anyone coming across this issue: we recommend using [Lineax](https://github.com/google/lineax) instead.
This solves the various issues discussed here.
